### PR TITLE
869 CSV Import service

### PR DESCRIPTION
**Pull request by jmilljr24**
#869

# ✍️ Description

Import Service for CSV files uploaded by organizations. Currently implemented base off Google Form's CSV files.

# 📷 Screenshots/Demos

### CODE REVIEW

File: app/services/organizations/csv_import_service.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        previous = FormSubmission.where(person: person, csv_timestamp: timestamp)
+        next unless previous.empty?
+
+        ActiveRecord::Base.transaction do
+          form_submission = FormSubmission.create(person: person, csv_timestamp: timestamp)
+          row.each do |col|
+            # skip Email and Timestamp col as they are saved on Form Submission
+            next if col[0] == "Email" || col[0] == "Timestamp"
+
+            FormAnswer.create(person: person, form_submission: form_submission,
```
Comment (kasugaijin) [inline app/services/organizations/csv_import_service.rb:30]: do we need to save the person on the FormAnswer? FormAnswer belongs_to FormSubmission, which belongs_to Person, and we can set FormAnswer belongs_to Person through FormSubmission. I think i most contexts we will be calling FormAnswer from FormSubmission rather than Person, because we will want to always be considering the FormAnswer context i.e. only give me those answers from the same submission.

**Comment by kasugaijin** [comment]
@jmilljr24 I think this is a good start!

File: app/services/organizations/csv_import_service.rb (hunk 1 above)
Comment (jmilljr24) [reply app/services/organizations/csv_import_service.rb:30]: Agreed. The old submitted answer had a user column so I did it as a placeholder without thinking through the associations. I'll adjust when the FormAnswer PR is complete so I get everything correct.

Comment (kasugaijin) [reply app/services/organizations/csv_import_service.rb:30]: Ah yes. I’ve also asked in the PR to name FormAnswer to remove user from that model association

**Comment by jmilljr24** [comment]
@kasugaijin This is updated to work with FormAnswer. 

Waiting on the email format that will be used in the DB and if we decided on a timestamp column.

File: test/services/organizations/csv_import_service_test.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
+require "test_helper"
+require "csv"
+
+class Organizations::CsvImportServiceTest < ActiveSupport::TestCase
```
Comment (kasugaijin) [inline test/services/organizations/csv_import_service_test.rb:4]: can we nest modules and classes from now on please :)

**Comment by kasugaijin** [comment]
Looking good @jmilljr24 
For the email, we are only collecting it on one model, User, and currently only on one form. What I suggest we do is have a before save hook. It will serve MVP fine and will do for a long time. I don't see us adding any new models that collect email any time soon.

```
class User < ApplicationRecord
  before_save :downcase_email

  private

  def downcase_email
    self.email = email.downcase if email.present?
  end
end
```

With respect to the CSV timestamp, I like your proposed approach to store this on the FormSubmission as `csv_timestamp`. 

Did you want to include both of those changes in this PR?

**Comment by jmilljr24** [comment]
> Looking good @jmilljr24 For the email, we are only collecting it on one model, User, and currently only on one form - adopter sign up. What I suggest we do is have a before save hook. It will serve MVP fine and will do for a long time. I don't see us adding any new models that collect email any time soon.
> 
> ```
> class User < ApplicationRecord
>   before_save :downcase_email
> 
>   private
> 
>   def downcase_email
>     self.email = email.downcase if email.present?
>   end
> end
> ```
> 
> With respect to the CSV timestamp, I like your proposed approach to store this on the FormSubmission as `csv_timestamp`.
> 
> Did you want to include both of those changes in this PR?

Yes. I can include the changes. Thanks!

**Review by kasugaijin — APPROVED** [review]
LGTM thanks @jmilljr24 
I was thinking...once we start actually using this in the wild, if there are nuances per third party service, we might want to write a parser specific to each service to handle its nuance. We could then include a dropdown select next to the form upload that specifies where the data came from. Then we can be sure to handle things neatly and be able to clearly support X services. It will avoid issues where a user uploads something from some obscure service, or worse, their own CSV, and wonders why it doesn't work as expected. Thoughts? Not something we have to do now, but I bet we'll run into this sooner than later.

**Comment by jmilljr24** [comment]
> LGTM thanks @jmilljr24 I was thinking...once we start actually using this in the wild, if there are nuances per third party service, we might want to write a parser specific to each service to handle its nuance. We could then include a dropdown select next to the form upload that specifies where the data came from. Then we can be sure to handle things neatly and be able to clearly support X services. It will avoid issues where a user uploads something from some obscure service, or worse, their own CSV, and wonders why it doesn't work as expected. Thoughts? Not something we have to do now, but I bet we'll run into this sooner than later.

Yes I had a similar thought about the dropdown. For the time being, we should definitely make a note on the upload page that lets the admin know this is only setup up for google forms right now. I also left the comment regarding the headers being specific to Google Forms for the next dev.

If there is any way we could test out one of the stakeholders csv files before MVP I would feel a bit better about this PR. There can be a lot a edge cases with csv's but I didn't want to add complexity without an actual need. I'm just not sure what we will come across with their files.

**Comment by kasugaijin** [comment]
Yes I agree it would be good to add a note re google forms to the upload form and also do some testing. I think once we are at MVP we can test with real data. I just hesitate to ask for all that private data right now. I think the next best thing we can do is make our own google form and do some testing. We can make an issue for this once the form input route, controller and view are in place.
